File: scrypt/pbkdf2hmacsha256.js
(function(){
	'use strict';
	
	var HEAP = new ArrayBuffer(4096 + 32 + 132)
		, HEAP_T = new Uint8Array(HEAP, 0, 4096)
		, HEAP_Ur = new Uint8Array(HEAP, 4096, 32)
		, HEAP_Ui_84 = new Uint8Array(HEAP, 4128, 84)
		, HEAP_Ui_132 = new Uint8Array(HEAP, 4128, 132);
	
	var hmacsha256 = this.hmacsha256;
	if (!hmacsha256 && (typeof require !== 'undefined')) var hmacsha256 = require('./hmacsha256');
	
	var intify = function(dest, offset, i) {
		dest[offset] = i >> 24;
		dest[offset + 1] = i >> 16;
		dest[offset + 2] = i >> 8;
		dest[offset + 3] = i;
	};
	
	var f = function(dest, offset, password, salt, blockIndex) {
		var Ur = HEAP_Ur
			, Ui = (salt.length == 128 ? HEAP_Ui_132 : HEAP_Ui_84)
			, i;
		
		// Clear out Ur
		i = 32;
		while (i--) {
			Ur[i] = 0;
		}
		
		// Copy
		i = salt.length;
		while (i--) {
			Ui[i] = salt[i];
		}
		
		intify(Ui, salt.length, blockIndex);
		
		// This loop not required for our usage of scrypt
//		for (i = 0; i < count; i++) {
			hmacsha256(password, Ui);
			
			i = 32;
			while (i--) {
				Ur[i] ^= Ui[i];
			}
//		}
		
		// Copy
		i = 32;
		while (i--) {
			dest[i + offset] = Ur[i];
		}
	};
	
	var pbkdf2hmacsha256 = function(password, salt, dkLen, DK) {
		var l = (dkLen > 32 ? dkLen : 32)
			, T = HEAP_T
			, i
			, tiOffset = 0;
		
		for (i = 1; i <= l; i++) {
			f(T, tiOffset, password, salt, i);
			tiOffset += 32;
		}
		
		// This check is not applicable to scrypt
		// if (r < 32) {
			// Copy
			i = dkLen;
			while (i--) {
				DK[i] = T[i];
			}
		// }
		// return T;
	};
	
	if (typeof exports !== 'undefined') {
		module.exports = pbkdf2hmacsha256;
	} else {
		this.pbkdf2hmacsha256 = pbkdf2hmacsha256;
	}
}).call(this);

File: scrypt/scrypt/pbkdf2hmacsha256.cc
#include <string.h>

#include "hmacsha256.hh"
#include "pbkdf2hmacsha256.hh"

void intify(unsigned char *destination, unsigned int offset, unsigned int i) {
	destination[offset] = i >> 24;
	destination[offset + 1] = i >> 16;
	destination[offset + 2] = i >> 8;
	destination[offset + 3] = i;
}

void f(unsigned char *destination, unsigned int offset, unsigned char *pass, unsigned int pl, unsigned char *salt, unsigned int sl, unsigned int index) {
	unsigned int i;
	unsigned char ur[32];
	unsigned char ui[132];
	
	// Reset ur
	memset(ur, 0, 32);
	
	// Copy salt into ui
	memcpy(ui, salt, sl);
	
	intify(ui, sl, index);
	
	HMACSHA256(pass, ui, (sl == 128 ? 132 : 84), ui);
	
	i = 32;
	while (i--) {
		ur[i] ^= ui[i];
	}
	
	// Copy ur into destination with offset
	memcpy(destination + offset, ur, 32);
}

void PBKDF2HMACSHA256(unsigned char *pass, unsigned int pl, unsigned char *salt, unsigned int sl, unsigned char *out, unsigned int ol) {
	unsigned int l = 32;
	if (ol > 32) l = ol;
	
	unsigned int i;
	unsigned int ti = 0;
	unsigned char t[4096];
	
	for (i = 1; i <= l; i += 1) {
		f(t, ti, pass, pl, salt, sl, i);
		ti += 32;
	}
	
	// Copy t into out
	memcpy(out, t, ol);
}
